### PR TITLE
Fix sram accesses.

### PR DESCRIPTION
**Pull request by shuffle2**
The meat of the PR is just being a little less dumb (but still dumb) about how to interpret the address sent to the "ipl" device.

Notably, rtc value is just maintained in the sram region, so I merged it into `struct Sram`.

### CODE REVIEW

File: Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp [hunk 1]
```diff
@@ -153,12 +147,11 @@ CEXIIPL::~CEXIIPL()
 }
 void CEXIIPL::DoState(PointerWrap& p)
 {
-  p.Do(m_RTC);
-  p.Do(m_uPosition);
-  p.Do(m_uAddress);
-  p.Do(m_uRWOffset);
+  p.Do(m_command);
```
Comment (shuffle2) [inline Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp]: oh i just noticed i forgot to add-back saving of rtc...

**Comment by degasus** [comment]
Based on the title, I assume there is a functional change in the patch? Inlining it with such a big cleanup is hard to review :/

File: Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp [hunk 2]
```diff
@@ -347,68 +290,88 @@ void CEXIIPL::TransferByte(u8& _uByte)
...
+      if (m_command.is_write())
+        g_SRAM.raw[dev_addr] = data;
       else
+        data = g_SRAM.raw[dev_addr];
+    }
+    else if (IN_RANGE(UART))
+    {
+      switch (DEV_ADDR(UART))
+      {
+      case 0:
```
Comment (sepalani) [inline Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp]: Shouldn't constants be defined for these UART magics.

File: Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp [hunk 3]
```diff
@@ -347,68 +290,88 @@ void CEXIIPL::TransferByte(u8& _uByte)
...
+      // Wii only RTC flags... afaik only the Wii Menu initializes it
+      // Seems to be 4bytes at dev_addr 0x20
+    }
+    else if (IN_RANGE(EUART))
+    {
+      switch (DEV_ADDR(EUART))
       {
-        NOTICE_LOG(OSREPORT, "EXI IPL-DEV: %s %x at %08x", IsWriteCommand() ? "write" : "read",
-                   _uByte, m_uAddress);
+      case 0:
```
Comment (sepalani) [inline Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp]: Ditto for EUART.

File: Source/Core/Core/HW/Sram.cpp [hunk 4]
```diff
@@ -11,12 +11,12 @@
 #include "Core/ConfigManager.h"
 
 // english
-const SRAM sram_dump = {{0x00, 0x2C, 0xFF, 0xD0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-                         0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x2C, 0x44, 0x4F,
-                         0x4C, 0x50, 0x48, 0x49, 0x4E, 0x53, 0x4C, 0x4F, 0x54, 0x41, 0x44,
-                         0x4F, 0x4C, 0x50, 0x48, 0x49, 0x4E, 0x53, 0x4C, 0x4F, 0x54, 0x42,
-                         0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-                         0x00, 0x00, 0x00, 0x6E, 0x6D, 0x00, 0x00, 0x00, 0x00}};
+const Sram sram_dump = {{0,    0,    0,    0,    0x00, 0x2C, 0xFF, 0xD0, 0x00, 0x00, 0x00, 0x00,
```
Comment (sepalani) [inline Source/Core/Core/HW/Sram.cpp]: This should probably be heavily detailed. How was it dump? From which region, system, version, etc.?

File: Source/Core/Core/HW/Sram.h [hunk 5]
```diff
@@ -34,55 +34,102 @@ distribution.
 -------------------------------------------------------------*/
 #pragma once
 
+#include <array>
 #include "Common/CommonTypes.h"
 
+typedef std::array<u8, 12> CardFlashId;
```
Comment (sepalani) [inline Source/Core/Core/HW/Sram.h]: ```using```?

File: Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp [hunk 6]
```diff
@@ -347,68 +290,88 @@ void CEXIIPL::TransferByte(u8& _uByte)
...
       }
-      break;
+    }
+    else
+    {
+      NOTICE_LOG(EXPANSIONINTERFACE, "IPL-DEV Accessing unknown device");
     }
 
-    m_uRWOffset++;
+#undef DEV_ADDR_CURSOR
```
Comment (sepalani) [inline Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp]: Not sure about the use of defines instead of functions, here. Is there a noticeable speed up?

**Comment by BhaaLseN** [comment]
Any chance you could split that commit into multiple ones? Reviewing is a pain with this big-bang-there-you-go commit... :/

File: Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp (hunk 2 above)
Comment (shuffle2) [reply Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp]: they are just unknown register indices

File: Source/Core/Core/HW/Sram.cpp (hunk 4 above)
Comment (shuffle2) [reply Source/Core/Core/HW/Sram.cpp]: There is only a change here because of clang-format

File: Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp (hunk 6 above)
Comment (shuffle2) [reply Source/Core/Core/HW/EXI/EXI_DeviceIPL.cpp]: I don't see a way to replace this with non-macro.

File: Source/Core/Core/HW/Sram.h (hunk 5 above)
Comment (shuffle2) [reply Source/Core/Core/HW/Sram.h]: i don't see the point, but i've changed it to `using` anyways...

**Comment by JosJuice** [comment]
Please don't self-merge PRs this complex without getting an LGTM.

**Comment by shuffle2** [comment]
this PR was not complex

**Comment by shuffle2** [comment]
now we wait another year for the next fix to merge :)

**Comment by weihuoya** [comment]
ok, I'm waiting.

**Comment by LuismaSP89** [comment]
> 
> 
> now we wait another year for the next fix to merge :)

No please xD, no way. I´ll test a thousand times if you need, but don´t wait so long...

**Pull request merged**